securedrop/qa_loader.py

```diff
@@ -0,0 +1,223 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+import os
+import random
+import string
+import sys
+
+from argparse import ArgumentParser
+from datetime import datetime
+from flask import current_app
+from os import path
+
+from db import db
+from journalist_app import create_app
+from models import (Journalist, Source, Submission, SourceStar, Reply,
+                    JournalistLoginAttempt)
+from sdconfig import config as sdconfig
+
+random.seed('~(=^–^)')  # mrow?
+
+JOURNALIST_COUNT = 10
+SOURCE_COUNT = 50
+
+
+def random_bool():
+    return bool(random.getrandbits(1))
+
+
+def random_chars(len, nullable, chars=string.printable):
+    if nullable and random_bool():
+        return None
+    else:
+        return ''.join([random.choice(chars) for _ in range(len)])
+
+
+def bool_or_none():
+    return random.choice([True, False, None])
+
+
+def random_datetime(nullable):
+    if nullable and random_bool():
+        return None
+    else:
+        return datetime(
+            year=random.randint(1, 9999),
+            month=random.randint(1, 12),
+            day=random.randint(1, 28),
+            hour=random.randint(0, 23),
+            minute=random.randint(0, 59),
+            second=random.randint(0, 59),
+            microsecond=random.randint(0, 1000),
+        )
+
+
+def new_journalist():
+    # Make a diceware-like password
+    pw = ' '.join([random_chars(3, nullable=False) for _ in range(7)])
+    journalist = Journalist(random_chars(random.randint(3, 32),
+                                         nullable=False),
+                            pw,
+                            random_bool())
+    journalist.is_admin = bool_or_none()
+
+    journalist.is_totp = bool_or_none()
+    journalist.hotp_counter = (random.randint(-1000, 1000)
+                               if random_bool() else None)
+    journalist.created_on = random_datetime(nullable=True)
+    journalist.last_access = random_datetime(nullable=True)
+
+    db.session.add(journalist)
+
+
+def new_source():
+    fid_len = random.randint(0, 32)
```

Comment:
Note that we also want to avoid very small non-zero `fid_len` as there is a high probability of hitting `IntegrityError` for them also (I can attest for this random seed `fid_len = random.randint(4, 32)` worked no problem)

Reply:
I included empty FID's because that's allowed by the DB and could (but very, very likely won't) cause problems during future migrations.